Add proration modes to post to backend

swehner opened the pull request:
Add missing proration modes and forward them to the Backend when posting receipts

### Motivation
For supporting proration modes we need to send them to the Backend. Added extra parameter in POST receipt.
Added all proration modes in GoogleProrationModes

### Description
* Use GoogleProrationMode internally instead of an enum
* Add an interface that Amazon could potentially implement in the future

MarkVillacampa commented on public/src/main/java/com/revenuecat/purchases/models/StoreTransaction.kt:
```diff
@@ -92,7 +92,9 @@ data class StoreTransaction(
-    val subscriptionOptionId: String?
+    val subscriptionOptionId: String?,
+
+    val prorationMode: String?
```
Should we add some comment here? Seems like all other fields have some documentation.

swehner replied:
Added a small note! Thanks!

MarkVillacampa reviewed (approved):
Looks good 👍

greenietea commented:
I expected to see the `proration_mode` param to be sent to the backend but it's missing?
![image](https://user-images.githubusercontent.com/4013719/233444992-0a695296-ef4b-4878-9d65-1efd9a497670.png)

swehner commented:
> I expected to see the `proration_mode` param to be sent to the backend but it's not there?
> 
> ![image](https://user-images.githubusercontent.com/4013719/233444992-0a695296-ef4b-4878-9d65-1efd9a497670.png)

You're totally right - thanks for double checking it. I must have missed a line in the latest refactor and screwed it up.
Added it back and added a test.

![Screenshot 2023-04-20 at 22 11 51](https://user-images.githubusercontent.com/273236/233480360-7a508dc6-7ea8-43e9-97b1-41fc9c94c21a.png)

tonidero commented on api-tester/src/main/java/com/revenuecat/apitester/kotlin/GoogleProrationModeAPI.kt:
```diff
@@ -7,7 +7,10 @@ private class GoogleProrationModeAPI {
+            GoogleProrationMode.IMMEDIATE_WITH_TIME_PRORATION,
+            GoogleProrationMode.DEFERRED,
+            GoogleProrationMode.IMMEDIATE_AND_CHARGE_FULL_PRICE,
+            GoogleProrationMode.IMMEDIATE_AND_CHARGE_PRORATED_PRICE -> {}
```
We need to add these to the java version of these API tests as well

swehner replied:
Done!

tonidero commented on feature/google/src/main/java/com/revenuecat/purchases/google/PurchaseContext.kt:
```diff
@@ -0,0 +1,11 @@
+class PurchaseContext(
+    val productType: ProductType,
+    val presentedOffering: String?,
+    val subscriptionOptionSelected: String?,
```
Should we rename these two to `presentedOfferingId` and `selectedSubscriptionOptionId`?

swehner replied:
Sure

tonidero commented on feature/google/src/main/java/com/revenuecat/purchases/google/BillingWrapper.kt:
```diff
@@ -78,9 +79,7 @@ class BillingWrapper(
-    private val productTypes = mutableMapOf<String, ProductType>()
-    private val presentedOfferingsByProductIdentifier = mutableMapOf<String, String?>()
-    private val subscriptionOptionSelectedByProductIdentifier = mutableMapOf<String, String?>()
+    private val purchaseContext = mutableMapOf<String, PurchaseContext>()
```
❤️

tonidero commented on feature/google/src/main/java/com/revenuecat/purchases/google/BillingFlowParamsExtensions.kt:
```diff
@@ -2,12 +2,13 @@ package com.revenuecat.purchases.google
         setOldPurchaseToken(replaceProductInfo.oldPurchase.purchaseToken)
         replaceProductInfo.prorationMode?.let {
-            setReplaceProrationMode(it)
+            setReplaceProrationMode((it as GoogleProrationMode).playBillingClientMode)
```
This could potentially crash if for some reason we passed a non-GoogleProrationMode here and try to cast it... We could do something like:
```
(replaceProductInfo.prorationMode?.let {
   val googleProrationMode = it as? GoogleProrationMode
   if (googleProrationMode == null) {
      errorLog("Got non-Google proration mode")
   } else {
      setReplaceProrationMode(googleProrationMode.playBillingClientMode)
   }
}
```

swehner replied:
What's the issue with actually crashing? This is a programming error... and a pretty bad one. This would mean you're using the Google implementation and suddenly passing in Amazon objects, so I'd be fine with it crashing. IMO this would be even better than just logging an error that could go undetected.

tonidero replied:
I kind of agree with you that crashes are much more visible and could help detect issues sooner. However, I believe the policy we've been following in the SDKs is that we really don't want to cause crashes in our SDK so we always program defensively and try to fail gracefully. This could be a discussion for the whole team though. CC @aboedo, In case you have any thoughts.

tonidero replied:
Note that we are crashing devs apps here, and we're not sure in what situation their app might be, or how they would like to handle. That's the main reason we try to avoid crashing whenever possible.

tonidero commented on feature/google/src/main/java/com/revenuecat/purchases/google/BillingWrapper.kt:
```diff
@@ -743,14 +745,14 @@ class BillingWrapper(
+                        context.productType,
+                        context.presentedOffering,
+                        context.subscriptionOptionSelected,
+                        context.prorationMode,
```
Wonder if we could just make this `toStoreTransaction` receive a `PurchaseContext` object?

swehner replied:
I added an overload with it, but there are a lot of use cases that don't take a PurchaseContext, or mix from different sources, so I'd still leave the other creator.

tonidero commented on public/src/main/java/com/revenuecat/purchases/models/StoreTransaction.kt:
```diff
@@ -92,7 +92,12 @@ data class StoreTransaction(
+    val subscriptionOptionId: String?,
+
+    /**
+     * The proration_mode string to be sent to the backend.
```
Hmm this seems not that useful for developers? Not sure if they would need to get it from the SDK for something... Maybe something like:
"The prorationMode used to perform the upgrade/downgrade of this purchase. Null if it was not an upgrade/downgrade or if the purchase was restored. This is not available for Amazon purchases."

swehner replied:
Makes sense, I wasn't aware whether this is exposed to the app developers. Added the doc and additionally added the fact that it's the name of GoogleProrationMode enum

tonidero commented on public/src/main/java/com/revenuecat/purchases/models/StoreTransaction.kt:
```diff
@@ -92,7 +92,12 @@ data class StoreTransaction(
+    /**
+     * The proration_mode string to be sent to the backend.
+     */
+    val prorationMode: String?
```
Considering we are exposing the `ProrationMode` and `GoogleProrationMode` classes as well, any reason we don't use them here?

swehner replied:
Right... initially I didn't because of where I had put the `ProrationMode` interface. But I'll just change it.

tonidero reviewed (commented):
Looking good! Left a few comments we should follow through though.

tonidero commented on public/src/main/java/com/revenuecat/purchases/ProrationMode.kt:
```diff
@@ -0,0 +1,5 @@
+package com.revenuecat.purchases
+
+interface ProrationMode {
+    val name: String
```
Since this is part of the public API, it should have docs as well.

tonidero replied:
Also, I know this interface was added to eventually support Amazon if/when they add this functionality. Do we have any news of whether they are planning to add this in the medium term? Just wondering if we should remove the complexity of the new interface/downcasting until we have some clear news of that functionality being supported there... Even if it's not so future-proof

swehner replied:
I'd keep it... if it's part of the public/common interface we shouldn't have anything platform specific. And I think for the devs this is write-only, so there shouldn't be a major problem. All casts are done internally in our backend and should be rather safe. (Although I'll add the checks like you suggested).

tonidero replied:
The `PurchaseParams` interface already uses `GoogleProrationMode` directly... But well, I think it's ok

swehner replied:
But that's fine, right? It's easier for the user to just specify the Google type, and if Amazon supports it at some point, we'll add a specific method to set the amazon proration type.
Internally we could either keep it the same, or as it is planned here, we just keep the interface that the backend needs - which is just a string for what to send in.
There are only a few places that make the downcast, and those should be fine because you should always get the right object.
Actually in this case we should check that you can't use the googleProrationMode in the builder if the SDK is configured as Amazon.

tonidero replied:
Yeah, I think this is fine, and definitely makes the API simpler to use. As for adding the check, we could certainly at least log some error if they try to do that... But not needed for this PR I think.

swehner commented on .idea/codeStyles/Project.xml:
```diff
@@ -8,8 +8,6 @@
         </value>
       </option>
-      <option name="NAME_COUNT_TO_USE_STAR_IMPORT" value="2147483647" />
-      <option name="NAME_COUNT_TO_USE_STAR_IMPORT_FOR_MEMBERS" value="2147483647" />
```
Nope, that was an accidental add... will remove. Thanks for noticing

tonidero commented on public/src/main/java/com/revenuecat/purchases/models/GoogleProrationMode.kt:
```diff
@@ -25,5 +30,64 @@ enum class GoogleProrationMode(@BillingFlowParams.ProrationMode val playBillingC
+     * so he is charged the difference of $0.50 for his new subscription.
+     * On May 1st, Samwise is charged $36 for his new subscription tier and another $36 on May 1 of each year following.
+     */
+    IMMEDIATE_AND_CHARGE_PRORATED_PRICE(BillingFlowParams.ProrationMode.IMMEDIATE_AND_CHARGE_PRORATED_PRICE);
```
So I might be wrong but I think we didn't add these before because they weren't fully supported? (I might be wrong) Just want to make sure we want to allow users to use these.

swehner replied:
We're working on supporting them in the BE

tonidero commented on api-tester/src/main/java/com/revenuecat/apitester/kotlin/StoreTransactionAPI.kt:
```diff
@@ -24,6 +24,7 @@ private class StoreTransactionAPI {
             val su1: String? = storeUserID
             val purchaseType: PurchaseType = purchaseType
             val subscriptionOptionId = subscriptionOptionId
+            val prorationMode = prorationMode
```
Oh I just noticed we are not specifying the type here, so this wouldn't actually catch issues with the type. This and the above should have the type, like this:
`val prorationMode: ProrationMode? = prorationMode`

tonidero commented on public/src/main/java/com/revenuecat/purchases/models/GoogleProrationMode.kt:
```diff
@@ -1,14 +1,19 @@
-enum class GoogleProrationMode(@BillingFlowParams.ProrationMode val playBillingClientMode: Int) {
+enum class GoogleProrationMode(
+    @BillingFlowParams.ProrationMode val playBillingClientMode: Int
+) : com.revenuecat.purchases.ProrationMode {
```
We could import this proration mode with an alias like 
```
import com.revenuecat.purchases.ProrationMode as RCProrationMode
```
NABD though.

tonidero commented on public/src/main/java/com/revenuecat/purchases/models/StoreTransaction.kt:
```diff
@@ -92,7 +93,14 @@ data class StoreTransaction(
+    val subscriptionOptionId: String?,
+
+    /**
+     * The name of the prorationMode used to perform the upgrade/downgrade of this purchase.
```
Now it's not the name, so this should be `The prorationMode used to...`

tonidero reviewed (approved):
Nothing blocking, looks good!

tonidero commented:
Oh looks like tests failed now.
